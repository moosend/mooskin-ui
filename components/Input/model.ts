import { IInputCallbackData } from '../_utils/types/commonTypes';
import { IBoxComponentProps, IInputBoxComponentProps } from '../Box/model';

export interface IInputContainerComponentProps extends IBoxComponentProps {
	/** provide to make the input field disabled */
	disabled?: boolean;

	/** value to be passed down to the input */
	value?: string | number;

	/** what data is being used, helps whn extracting user input, you know on what field changes are made */
	dataLabel?: string;

	/** callback that is called when the input changes */
	onChangeInput?: (e: React.ChangeEvent<HTMLInputElement>, data: IInputCallbackData) => void;
}

export interface IInputComponentProps extends IInputBoxComponentProps {
	/** provide to make the input field disabled */
	disabled?: boolean;

	/** what data is being used, helps whn extracting user input, you know on what field changes are made */
	dataLabel?: string;

	/** min and max values for number values */
	minmax?: number[];

	/** numberType */
	numberType?: 'integer' | 'float';

	/** whether its wrapped by its container */
	wrapped?: boolean;
}

export interface IInputEmojiComponentProps extends IBoxComponentProps {
	/** onclick event handler when clicking on the emoji */
	onChangeEmoji?: (value: string) => void;
}

export interface IInputListComponentProps extends IBoxComponentProps {
	/** icon to be shown for toggling the dropdown */
	icon: string;

	iconProps?: IBoxComponentProps;

	wrapperProps?: IBoxComponentProps;
}

export interface IInputOptionComponentProps extends IBoxComponentProps {
	/** value of the clicked element */
	value: string;
}
