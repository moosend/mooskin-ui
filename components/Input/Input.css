@import '../_utils/globals/variables.css';
@import '../_utils/globals/screens.css';
@import url('https://fonts.googleapis.com/icon?family=Material+Icons');

.inputContainer{
    display: flex;
    /* flex: 1; */
    box-sizing: border-box;
    /* align-items: center; */
}

.column {
    flex-direction: column;
    align-items: stretch;
}

.input{
    outline: none;
    display: flex;
    flex: 1;
    font-family: Arial,sans-serif;
    border: 0;
    color: var(--inputFontColor);
    /* font-size: 16px; */
    font-size: 1rem; 
    transition: 0.2s;
    padding: 15px;
    background: transparent;
}

.innerDiv{
    display: flex;
    flex: 1;
    align-items: center;
    justify-content: center;
    position: relative;
    background: #efefef;
    border-radius: 8px;
    border-radius: 8px;
}

.reverse{
    flex-direction: row-reverse;
}

.inputLeft{
    border-top-left-radius: 8px;
    border-bottom-left-radius: 8px;
    padding: 15px 5px 15px 15px;
}

.inputRight{
    border-top-right-radius: 8px;
    border-bottom-right-radius: 8px;
    padding: 15px 15px 15px 5px;
}

.disabledInput{
    cursor:not-allowed;
    color: #bfbfbf;
    background: #f6f6f6;
}

.input::placeholder {
    color: var(--inputPlaceholder);
}

.iconContainer{
    display: flex;
    padding: 15px 15px 15px 0;
}

.icon{
    color: var(--inputPlaceholder);
    font-size: 20px;
    font-family: 'Material Icons';
    background: transparent;
}

.iconError{
    color: #ec3232;
}

.iconSuccess{
    color: #00cc44;
}

.input:focus {
    outline: none;
    color:#333333;
}

.inputLabel {
    color: #6b6b6b;
     /* font-size: 13px;  */
    font-size: .92857142857rem; 
    cursor: pointer;
    flex-basis: 150px;
    padding-top: 15px;
}

.topLabel {
    flex-basis: auto;
    padding-top: 0px;
    margin-left: 3px;
}

.description{
    left: 5px;
    display: block;
    color: var(--italicDescription);
    font-size: .78571428571rem;
    margin-top: 10px;
}

.success{
    box-shadow: inset 0px 0px 0px 2px #00cc44;
}

.error{
    box-shadow: inset 0px 0px 0px 2px #ec3232;
}

.descError{
    color: #ec3232;
}

.descSuccess{
    color: #00cc44;
}

.copyButton {
    margin-left: 15px;
}

.dropDown {
    position: absolute;
    top: 25px;
    right: -7px;
    height: auto;
    background: #fff;
    border-radius: 5px;
    z-index: 4;
    display: none;
    box-shadow: 0 1px 5px rgba(0,0,0,.32);
}

.dropDownArrow {
    width: 0;
    height: 0;
    border-left: 5px solid transparent;
    border-right: 5px solid transparent;
    border-bottom: 5px solid #F5F7F9;
    top: -5px;
    right: 10px;
    z-index: 5;
    position: absolute;
}

.dropDownLabel {
    padding: 5px;
    border-bottom: 1px solid #f1f1f1;
    background-color: #F5F7F9;
    text-align: center;
    border-top-left-radius: 5px;
    border-top-right-radius: 5px;
}

.dropDownContent {
    max-height: 400px;
    overflow-y: auto;
    width: 250px;
    padding: 5px;
    border-bottom-left-radius: 5px;
    border-bottom-right-radius: 5px;
}

.overlay {
    position: fixed;
    top: 0px;
    left: 0px;
    right: 0px;
    bottom: 0px;
}

.tag {
    padding-top: 3px;
    padding-bottom: 3px;
}

.dropDownContent > .tag:not(:last-child) {
    border-bottom: 1px solid #f1f1f1;
}

.tag:hover {
    background-color: #f9f9f9;
    cursor: pointer;
}

.dropDownIcon{
    padding-left: 10px;
}

.dropDownIcon:hover{
    cursor: pointer;
}

:global(.emoji-dialog .emoji-row .emoji){
    width: 35px;
    height: 35px;
    padding: 5px;
}

:global(.emoji-dialog .emoji-row .emoji):hover{
    cursor: pointer;
}

:global(.emoji-dialog .emojione){
    width: 25px;
}

:global(.emoji-dialog .emoji-dialog-header li img):hover{
    cursor: pointer;
    filter: grayscale(0);
}

:global(.emoji-dialog .emoji-dialog-header li img){
    height: 26px;
    width: 26px;
}

:global(.emoji-dialog .emoji-dialog-header .emoji){
    height: 26px;
    width: 26px;
}

:global(.emoji-dialog){
    width: 310px;
    height: 285px;
}

@media (--xSmall){
    .inputContainer{
        flex-direction: column;
        align-items: stretch;
    }

    .innerContainer {
        flex-direction: column;
        align-items: stretch;
    }
    
    .copyButton{
        margin: 5px 0 0 0;
    }
    
    .inputLabel{
        flex-basis: auto;
        padding-top: 0px;
        margin-left: 3px;
    }

}
