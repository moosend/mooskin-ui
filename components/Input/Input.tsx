import * as React from 'react';

import EmojiPicker from 'emojione-picker';
import 'emojione-picker/css/picker.css';

import styles from './Input.css';

import ClipboardButton from '../ClipboardButton';

import emojiIcon from '../../assets/images/editor/emoji.png';
import hashtag from '../../assets/images/editor/hashtag.png';

import {IInputCallbackData, IValidationCallbackData} from '../_utils/types/commonTypes';

export interface IProps {

    /** override input id */
    id?: string;

    /** provide to make the input field disabled */
    disabled?: boolean;

    /** provide to make the input field required */
    required?: boolean;

    /** override input type */
    type?: string;

    /** override input name */
    name?: string;

    /** override input value */
    value: string | number;

    /** min and max values for number values */
    minmax?: number[];

    /** override input placeholder */
    placeholder?: string;

    /** override input minlength */
    minlength?: number;

    /** override input maxlength */
    maxlength?: number;

    /** input label */
    label?: string;

    /** sets emoji dropdown to the input for adding emojis */
    emoji?: boolean;

    /** add personalization tags dropdown */
    personalizationTags?: Array<{value: string, label: string}>;

    /** close personalization tags dropdown when a tag is added */
    closeOnTagAdd?: boolean;

    /** spacing between label and input */
    labelWidth?: number;

    /** place label on top of Input */
    labelTop?: boolean;

    /** toggle autocomplete specified input */
    autocomplete?: boolean;

    /** autofocus specified input */
    autofocus?: boolean;

    /** icon to be placed near input */
    icon?: string;

    /** position of the icon */
    iconPosition?: string;

    /** override icon styles */
    iconStyle?: React.CSSProperties;

    /** override icon class */
    iconClass?: string;

    /** override input styles */
    style?: React.CSSProperties;

    /** override input class */
    className?: string;

    /** extra html attributes */
    extraHtmlAttr?: {[key: string]: any};

    /** type of input should be a div instead of input */
    divType?: boolean;

    /** input description (small italic bottom) */
    description?: string;

    /** status of the input, error or success */
    status?: 'error' | 'success';

    /** adds clipboardButton to the input component and assigns a label */
    clipboardButton?: string;

    /** clipboardButton callback function when clicked */
    onClipboardButtonClick?: (e: React.MouseEvent<HTMLButtonElement>, data: IInputCallbackData) => void;

    /** validate function */
    validate?: (data: IValidationCallbackData) => boolean;

    /** what data is being used, helps whn extracting user input, you know on what field changes are made */
    dataLabel?: string;

    /** callback that is called when the input changes */
    onChange?: (e: React.ChangeEvent<HTMLInputElement>, data: IInputCallbackData) => void;

}

export interface IInputState{
    activeDropDown: number;
}

class Input extends React.Component<IProps, IInputState> {

    static defaultProps = {
        className: '',
        closeOnTagAdd: true,
        iconPosition: 'right',
        style: {}
    };

    id: string;
    input: any;

    constructor(props: IProps){
        super(props);

        this.id = this.props.id || this.generateId();

        this.state = {
            activeDropDown: -1
        };

    }

    render(){

        const {
            style,
            className,
            label,
            clipboardButton,
            labelTop,
        } = this.props;

        const spacing = !this.props.labelWidth ? {} : {flexBasis: `${this.props.labelWidth}px`};

        const labelPos = labelTop ? styles.column : '';
        const topLabel = labelTop ? styles.topLabel : '';
        const inputClasses = `${styles.inputLabel} ${topLabel}`;

        return (
            <div className={`input-component ${styles.inputContainer} ${labelPos} ${className}`} style={style}>
                {label && <label className={inputClasses} style={spacing} htmlFor={this.id}>{label}</label>}
                {this.renderInput()}
                {clipboardButton && this.getClipboardButton()}
                {this.state.activeDropDown !== -1 && <div onClick={this.removeDropDown} className={styles.overlay} />}
            </div>
        );
    }

    renderInput = () => {
        const {
            autofocus,
            disabled,
            divType,
            maxlength,
            minlength,
            placeholder,
            required,
            iconPosition,
            description,
            icon,
            type
        } = this.props;

        const autocomplete = !this.props.autocomplete ? 'off' : 'on';
        const disabledInput = disabled ? styles.disabledInput : '';

        const value = type === 'number' ? this.getNumberValue(this.props.value as number) : this.props.value;

        if (divType){
            return (
                <div
                    ref={(input) => this.input = input}
                    // onChange={this.onChange}
                    id={this.id}
                    // value={this.props.value}
                    className={`input ${styles.input} ${disabledInput}`}
                    onBlur={this.validateOnBlur}
                    {...this.props.extraHtmlAttr}
                />
            );
        } else {
            const status = this.getStatus();
            const descStatus = this.getDescStatus();
            const reverse = iconPosition === 'left' && styles.reverse;
            return (
                <div style={{flex: 1}}>
                    <div className={`${styles.innerDiv} ${status} ${reverse} ${disabledInput}`}>
                        <input
                            ref={(input) => this.input = input}
                            onChange={this.onChange}
                            id={this.id}
                            type={type}
                            name={this.props.name}
                            value={value}
                            placeholder={placeholder}
                            minLength={minlength}
                            maxLength={maxlength}
                            required={required}
                            disabled={disabled}
                            className={styles.input}
                            autoFocus={autofocus}
                            autoComplete={autocomplete}
                            onBlur={this.validateOnBlur}
                            {...this.props.extraHtmlAttr}
                        />
                        <div className={styles.iconContainer}>
                            {this.getDropDown()}
                            {icon && this.getIcon()}
                        </div>
                    </div>
                    {description && <i className={`${styles.description} ${descStatus}`}>{description}</i>}
                </div>
            );
        }
    }

    onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const returnValue = this.props.type === 'number' ? this.getNumberValue(e.target.valueAsNumber) : e.target.value;
        !this.props.disabled &&
        this.props.onChange &&
        this.props.onChange(e, {value: returnValue, dataLabel: this.props.dataLabel});
        if (this.props.status){
            this.props.validate &&
            this.props.validate(
                {value: returnValue, dataLabel: this.props.dataLabel, required: this.props.required}
            );
        }
    }

    validateOnBlur = () => {
        this.props.validate &&
        this.props.validate({value: this.props.value, dataLabel: this.props.dataLabel, required: this.props.required});
    }

    getStatus = () => {
        const inputStatus = this.props.status && this.props.status;
        if (inputStatus){
            if (inputStatus === 'error'){
                return styles.error;
            } else if (inputStatus === 'success'){
                return styles.success;
            }
        }
    }

    getDescStatus = () => {
        const inputStatus = this.props.status && this.props.status;
        if (inputStatus){
            if (inputStatus === 'error'){
                return styles.descError;
            } else if (inputStatus === 'success'){
                return styles.descSuccess;
            }
        }
    }

    getNumberValue = (value: number) => {
        if (isNaN(value)){
            return 0;
        }
        if (this.props.minmax){
            let min;
            let max;
            if (this.props.minmax.length > 2) {
                throw new Error(`Prop 'minmax' has more values than expected!`);
            } else if (this.props.minmax.length === 0) {
                throw new Error(`Prop 'minmax' has no values!`);
            } else if (this.props.minmax.length === 1) {
                max = this.props.minmax && this.props.minmax[0];
            } else if (this.props.minmax.length === 2){
                min = this.props.minmax && this.props.minmax[0];
                max = this.props.minmax && this.props.minmax[1];
            }
            if (min && value < min){
                return min;
            } else if (max && value > max) {
                return max;
            } else {
                return value;
            }
        }
        return value;
    }

    generateId = () => {
        return Math.random().toString(36).substr(2, 10);
    }

    getIcon = () => {
        const iconFont = this.props.icon && this.getIconContent();
        const iconStatus = this.getIconStatus();
        const iconPadding = this.props.iconPosition === 'right' ? {paddingLeft: 10} : {paddingRight: 10};
        return (
            <div
                onClick={this.onIconClick}
                className={`${styles.icon} ${iconStatus} ${this.props.iconClass}`}
                style={{...iconPadding, ...this.props.iconStyle}}
            >
                {iconFont}
            </div>
        );
    }

    getIconStatus = () => {
        const inputStatus = this.props.status && this.props.status;
        if (inputStatus){
            if (inputStatus === 'error'){
                return styles.iconError;
            } else if (inputStatus === 'success'){
                return styles.iconSuccess;
            }
        }
    }

    getIconContent = () => {
        return this.props.icon ? this.props.icon.replace(/\s/g, '_') : '';
    }

    onIconClick = () => {
        this.input.focus();
    }

    getClipboardButton = () => {
        return (
            <ClipboardButton
                label={this.props.clipboardButton}
                value={this.props.value}
                className={styles.copyButton}
                style={{alignSelf: 'center'}}
                onClick={this.onClipboardButtonClick}
            />
        );
    }

    onClipboardButtonClick = (e: React.MouseEvent<HTMLButtonElement>) => {
        this.props.onClipboardButtonClick &&
        this.props.onClipboardButtonClick(e, {value: this.props.value, dataLabel: this.props.dataLabel});
    }

    getDropDown = () => {
        const dropdowns = [];
        let i = 0;
        this.props.personalizationTags &&
        dropdowns.push(this.getDropDownIcon(hashtag, i, 'personalization')) && (i = i + 1);
        this.props.emoji && dropdowns.push(this.getDropDownIcon(emojiIcon, i, 'emoji'));
        return dropdowns;
    }

    getDropDownIcon = (icon: string, i: number, type: string) => {
        const display = this.state.activeDropDown === i ? {display: 'block'} : {display: 'none'};
        return (
            <div
                className="dropdown-icon"
                key={i}
                style={{position: 'relative', display: 'flex', alignItems: 'center'}}
                onClick={() => this.onDropDownIconClick(i)}
            >
                <img className={styles.dropDownIcon} src={icon}/>
                {this.renderDropDown(display, type)}
            </div>
        );
    }

    onDropDownIconClick = (i: number) => {
        this.setState({activeDropDown: i});
    }

    removeDropDown = () => {
        this.setState({activeDropDown: -1});
    }

    renderDropDown = (display: React.CSSProperties, type: string) => {
        return (
            <div style={display} className={styles.dropDown}>
                <div className={styles.dropDownArrow} />
                {type === 'personalization' && this.getPersDropDown()}
                {type === 'emoji' && this.getEmojis()}
            </div>
        );
    }

    getPersDropDown = () => {
        return (
            <>
                <div className={styles.dropDownLabel}>Personalization Tags</div>
                <div className={styles.dropDownContent}>
                    {this.getPersonalizationTags()}
                </div>
            </>
        );
    }

    getPersonalizationTags = () => {
        return this.props.personalizationTags && this.props.personalizationTags.map((tag, i) => {
            return (
                <div key={i} onClick={(e) => this.addTag(e, tag.value)} className={styles.tag}>
                    {tag.label}
                </div>
            );
        });
    }

    addTag = (e: any, value: string) => {
        const finalValue = this.props.value ? this.props.value + value : value;
        this.props.onChange &&
        this.props.onChange(e, {value: finalValue, dataLabel: this.props.dataLabel});
        if (this.props.status){
            this.props.validate &&
            this.props.validate(
                {value: finalValue, dataLabel: this.props.dataLabel, required: this.props.required}
            );
        }
        setTimeout(() => {
            this.props.closeOnTagAdd && this.removeDropDown();
        }, 1);
    }

    getEmojis = () => {
        return <EmojiPicker onChange={this.onEmojiChange} />;
    }

    onEmojiChange = (data: any) => {
        const emoji = String.fromCodePoint(parseInt('0x' + data.unicode, 16));
        const finalValue = this.props.value ? this.props.value + emoji :  emoji;
        this.props.onChange &&
        this.props.onChange({} as any, {value: finalValue, dataLabel: this.props.dataLabel});
        if (this.props.status){
            this.props.validate &&
            this.props.validate(
                {value: finalValue, dataLabel: this.props.dataLabel, required: this.props.required}
            );
        }
    }

}

export default Input;
