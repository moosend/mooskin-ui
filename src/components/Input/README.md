# MooSkin UI - Input Component

The MooSkin Input Component works similar to the normal HTML `<input/>` element, but with pre-defined styling and attributes.

___

### Usage

To start using the Input Component first you have to Import it

```
Import {Input} from 'mooskin';
```

And then you can simply start using it by typing

```
<Input />
```

For easy use, the components are named similar to normal HTML components but with a capital first letter.

Like the `<input/>` element it will accept given attributes and render differently based on the given attributes

### Examples

Passing the placeholder attribute

```
<Input placeholder="username" />
```

or making it disabled with a value, in this case the proper disabled class will be loaded (ex. disabling cursor)

```
<Input value="This one is disabled" disabled />
```

or pass it a function for event handling

```
<Input onChange={yourFunc} />
```

or just give it a custom style

```
<Input style={yourStyle} />
```

## Supported attributes

* `id` - id of the element
* `type` - type of the input (eg. text)
* `name` - name of the input
* `value` - text value of input
* `placeholder` - backtext description of input
* `minlength` - min number of characters allowed
* `maxlength` - max number of characters allowed
* `disabled` - input field should be disabled
* `required` - input field is required
* `className` - css class
* `dataLabel` - label what kind of data 
* `style` - input field style
* `onChange` - callback to be triggered on input change

Allthough these attributes are supported, all of them are optional.

___


#### For more

[MooSkin-UI](https://github.com/moosend/mooskin-ui)