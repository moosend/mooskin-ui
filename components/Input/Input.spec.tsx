import * as React from 'react';
import Input from './Input';

import { mount, shallow } from 'enzyme';

describe('Input', () => {

    test('renders correctly', () => {
        const func = jest.fn();
        Date.now = jest.fn(() => 1482363367071);
        Math.random = jest.fn(() => 222333444555);

        const tree = shallow(
            <Input
                onChange={func}
                disabled
                className="myClass"
                style={{color: 'blue'}}
                id={'input1'}
                value="input value"
                placeholder="input placeholder"
                maxlength={50}
                minlength={10}
                name="input name"
                type="text"
                description="just some input"
                label="with a label ofc"
                labelWidth={30}
                autofocus
                autocomplete
                required
            />
        );
        expect(tree).toMatchSnapshot();
    });

    test('renders properly into dom and has Placeholder prop', () => {

        const component = shallow(<Input value="value" placeholder="username"/>);

        expect(component.find('input').prop('placeholder')).toContain('username');
    });

    test('renders an input with a "required" prop and minlength', () => {
        const component = shallow(<Input value="value" minlength={5} required/>);

        expect(component.find('input').prop('required')).toEqual(true);
        expect(component.find('input').prop('minLength')).toEqual(5);
    });

    test('renders an input with a passed value and maxlength', () => {
        const component = shallow(<Input value="random" maxlength={50}/>);

        expect(component.find('input').prop('value')).toEqual('random');
        expect(component.find('input').prop('maxLength')).toEqual(50);
    });

    test('renders an input with id and type', () => {

        const component = shallow(<Input value="value" type="text" id="1234"/>);

        expect(component.find('input').prop('id')).toEqual('1234');
        expect(component.find('input').prop('type')).toEqual('text');
    });

    test('renders an input with custom css class and style', () => {
        const component = shallow(<Input value="value" style={{color: 'blue'}} className="input-group"/>);

        expect(component.find('.inputContainer').hasClass('input-group')).toBe(true);
        expect(component.find('.inputContainer').prop('style')).toEqual({color: 'blue'});
    });

    test('onChange prop callback is called when a key is pressed', () => {
        const func = jest.fn();

        const component = shallow(<Input value="value" onChange={func}/>);

        component.find('input').simulate('change', { target: { value: 'text' }});
        expect(func).toHaveBeenCalled();
    });

    test('error classes get applied when input is not validated', () => {
        let inputValue = '';
        let status = null;

        const onChange = (e, data) => {
            inputValue = data.value;
        };

        const validate = (data) => {
            if (data.value < 5){
                status = 'error';
                return false;
            }
            return true;
        };

        const component = mount(<Input status={status} validate={validate} onChange={onChange} value={inputValue}/>);

        component.find('input').simulate('blur');
        expect(component.find('.error')).toBeTruthy();

        component.find('input').simulate('change', { target: { value: 'text' }});
        component.setProps({status: 'error'});
        expect(component.find('.error')).toBeTruthy();

        component.find('input').simulate('change', { target: { value: 'more text' }});
        expect(component.find('.error')).toBeFalsy;
    });

    test('Icon is valid when icon prop is passed', () => {

        const component = shallow(<Input value="value" icon="check" />);

        expect(component.find('.icon').length).toBe(1);
    });

    test('clipboardButton is available when the prop is passed', () => {

        const component = mount(<Input value="value" clipboardButton="Copy Text" />);

        expect(component.find('Button').length).toBe(1);
    });

    test('emoji dropdown is available when emoji prop is passed', () => {

        const component = mount(<Input value="value" emoji />);

        expect(component.find('.dropDown').length).toBe(1);

        expect(component.find('.dropDown').prop('style')).toEqual({display: 'none'});

        expect(component.state('activeDropDown')).toBe(-1);

        component.find('.dropDownIcon').simulate('click');

        expect(component.state('activeDropDown')).toBe(0);

        expect(component.find('.dropDown').prop('style')).toEqual({display: 'block'});
    });

    test('both dropdowns are available when relevant props are passed', () => {

        const pers = [
            {
                label: 'Unsibscribe Link',
                value: '#unsubscribeLink#'
            },
            {
                label: 'Account Email',
                value: '#account:email#'
            }
        ];

        const component = mount(<Input value="value" personalizationTags={pers} emoji />);

        expect(component.find('.dropDown').length).toBe(2);

        expect(component.find('.dropDown').first().prop('style')).toEqual({display: 'none'});
        expect(component.find('.dropDown').last().prop('style')).toEqual({display: 'none'});

        expect(component.state('activeDropDown')).toBe(-1);

        component.find('.dropDownIcon').first().simulate('click');

        expect(component.state('activeDropDown')).toBe(0);

        expect(component.find('.dropDown').first().prop('style')).toEqual({display: 'block'});
        expect(component.find('.dropDown').last().prop('style')).toEqual({display: 'none'});

        component.find('.dropDownIcon').last().simulate('click');

        expect(component.state('activeDropDown')).toBe(1);

        expect(component.find('.dropDown').first().prop('style')).toEqual({display: 'none'});
        expect(component.find('.dropDown').last().prop('style')).toEqual({display: 'block'});
    });

    test('onChange callback is called when a personalization tag is clicked', () => {

        const func = jest.fn();

        const pers = [
            {
                label: 'Unsibscribe Link',
                value: '#unsubscribeLink#'
            },
            {
                label: 'Account Email',
                value: '#account:email#'
            }
        ];

        const component = mount(<Input onChange={func} value="value" personalizationTags={pers} />);

        expect(component.find('.dropDown').length).toBe(1);

        component.find('.tag').first().simulate('click');

        expect(func).toHaveBeenCalled();

    });

    // test('assigns minmax prop correctly and returns the expected value', () => {

    //     let value = 5;

    //     const onChange = (e, data) => {
    //         value = data.value;
    //     };

    //     const component = mount(<Input onChange={onChange} type="number" value={value} minmax={[32]} />);

    //     component.find('input').simulate('change', { target: { valueAsNumber: 7 }});

    //     component.setProps({value});

    //     expect(component.prop('value')).toEqual(7);

    //     component.find('input').simulate('change', { target: { valueAsNumber: 35 }});

    //     component.setProps({value});

    //     expect(component.prop('value')).toEqual(32);

    // });

    // test('assigns minmax prop correctly and returns the expected value, chapter 2', () => {

    //     let value = '5';

    //     const onChange = (e, data) => {
    //         value = data.value;
    //     };

    //     const component = mount(<Input onChange={onChange} type="number" value={value} minmax={[32, 153]} />);

    //     component.find('input').simulate('change', { target: { valueAsNumber: 7 }});

    //     component.setProps({value});

    //     expect(component.prop('value')).toEqual(7);

    //     component.find('input').simulate('blur');

    //     component.setProps({value});

    //     expect(component.prop('value')).toEqual(32);

    //     component.find('input').simulate('change', { target: { valueAsNumber: 99 }});

    //     component.setProps({value});

    //     expect(component.prop('value')).toEqual(99);

    //     component.find('input').simulate('change', { target: { valueAsNumber: 170 }});

    //     component.setProps({value});

    //     expect(component.prop('value')).toEqual(153);

    // });

    test('close personalization tags dropdown when a personalization tag is added', () => {

        const pers = [
            {
                label: 'Unsibscribe Link',
                value: '#unsubscribeLink#'
            },
            {
                label: 'Account Email',
                value: '#account:email#'
            }
        ];

        const component = shallow(<Input value="value" personalizationTags={pers}/>);

        expect(component.find('.dropDown').prop('style')).toEqual({display: 'none'});

        component.find('.dropDownIcon').simulate('click');

        expect(component.find('.dropDown').prop('style')).toEqual({display: 'block'});

        component.find('.tag').first().simulate('click');

        setTimeout(() => {
            expect(component.find('.dropDown').prop('style')).toEqual({display: 'none'});
        }, 100);

    });

    test('renders custom dropdown when relevant prop is passed', () => {

        const customDropdown = {
            content: [
                <div key={1}>Content 1</div>,
                <div key={2}>Content 2</div>,
                <div key={3}>Content 3</div>,
                <div key={4}>Content 4</div>,
                <div key={5}>Content 5</div>
            ],
            icon: 'https://s1.piq.land/2015/10/05/Fix77XEm3n3EfO1dObGIEAnO_400x400.png',
            title: 'Custom Dropdown',
        };

        const component = shallow(<Input value="value" customDropdowns={customDropdown}/>);

        expect(component.find('.dropDown').length).toEqual(1);

        expect(component.find('.dropDownLabel').text()).toEqual(customDropdown.title);

        expect(component.find('.dropDownContent').prop('children')).toEqual(customDropdown.content);

    });

    test('renders multiple custom dropdowns when relevant prop is passed', () => {

        const multipleDropdowns = [
            {
                content: [
                    <div key={1}>Content 1</div>,
                    <div key={2}>Content 2</div>,
                    <div key={3}>Content 3</div>,
                    <div key={4}>Content 4</div>,
                    <div key={5}>Content 5</div>
                ],
                icon: 'https://s1.piq.land/2015/10/05/Fix77XEm3n3EfO1dObGIEAnO_400x400.png',
                title: 'Custom Dropdown 1',
            },
            {
                content: [
                    <div key={1}>Content 1</div>,
                    <div key={2}>Content 2</div>,
                    <div key={3}>Content 3</div>,
                    <div key={4}>Content 4</div>,
                    <div key={5}>Content 5</div>
                ],
                icon: 'https://doitgeekly.files.wordpress.com/2016/01/22c04c094e328d92fc7488fc6e2262b3.jpg',
                title: 'Custom Dropdown 2',
            },
            {
                content: [
                    <div key={1}>Content 1</div>,
                    <div key={2}>Content 2</div>,
                    <div key={3}>Content 3</div>,
                    <div key={4}>Content 4</div>,
                    <div key={5}>Content 5</div>
                ],
                icon: 'https://s1.piq.land/2016/02/17/kmeRltKJo5wELTWFipYHl5je_400x400.png',
                title: 'Custom Dropdown 3',
            }
        ];

        const component = shallow(<Input value="value" customDropdowns={multipleDropdowns}/>);

        expect(component.find('.dropDown').length).toEqual(3);

        multipleDropdowns.forEach((dropdown, i = 1) => {
            expect(component.find('.dropDownLabel').at(i).text()).toEqual(dropdown.title);

            expect(component.find('.dropDownContent').at(i).prop('children')).toEqual(dropdown.content);
        });

    });

    test('calls onClick function when icon is clicked', () => {

        const func = jest.fn();

        const multipleDropdowns = [
            {
                content: [
                    <div key={1}>Content 1</div>,
                    <div key={2}>Content 2</div>,
                    <div key={3}>Content 3</div>,
                    <div key={4}>Content 4</div>,
                    <div key={5}>Content 5</div>
                ],
                icon: 'https://s1.piq.land/2015/10/05/Fix77XEm3n3EfO1dObGIEAnO_400x400.png',
                onClick: func,
                title: 'Custom Dropdown 1',
            },
            {
                content: [
                    <div key={1}>Content 1</div>,
                    <div key={2}>Content 2</div>,
                    <div key={3}>Content 3</div>,
                    <div key={4}>Content 4</div>,
                    <div key={5}>Content 5</div>
                ],
                icon: 'https://doitgeekly.files.wordpress.com/2016/01/22c04c094e328d92fc7488fc6e2262b3.jpg',
                title: 'Custom Dropdown 2',
            },
            {
                content: [
                    <div key={1}>Content 1</div>,
                    <div key={2}>Content 2</div>,
                    <div key={3}>Content 3</div>,
                    <div key={4}>Content 4</div>,
                    <div key={5}>Content 5</div>
                ],
                icon: 'https://s1.piq.land/2016/02/17/kmeRltKJo5wELTWFipYHl5je_400x400.png',
                title: 'Custom Dropdown 3',
            }
        ];

        const component = shallow(<Input value="value" customDropdowns={multipleDropdowns}/>);

        expect(component.find('.dropDown').length).toEqual(3);

        component.find('.dropDownIcon').first().simulate('click');

        expect(func).toHaveBeenCalled();

    });

});
