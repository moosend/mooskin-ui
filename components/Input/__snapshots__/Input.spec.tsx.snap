// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
<div
  className="input-component inputContainer  myClass"
  style={
    Object {
      "color": "blue",
    }
  }
>
  <label
    className="inputLabel "
    htmlFor="input1"
    style={
      Object {
        "flexBasis": "30px",
      }
    }
  >
    with a label ofc
  </label>
  <div
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <div
      className="innerDiv undefined false disabledInput"
    >
      <input
        autoComplete="on"
        autoFocus={true}
        className="input"
        disabled={true}
        id="input1"
        maxLength={50}
        minLength={10}
        name="input name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="input placeholder"
        required={true}
        type="text"
        value="input value"
      />
      <div
        className="iconContainer"
      />
    </div>
    <i
      className="description undefined"
    >
      just some input
    </i>
  </div>
</div>
`;
